### DISCUSSION TITLE
OpenModelica FMU creation and usage

### DISCUSSION THREAD

**Discussion by jmythms** [ibpsa/project1-boptest#691] [Q&A - Software]
Hi,

I'm trying to change my OPTIMICA-based FMU creation workflow to use OpenModelica instead. The FMUs will be uploaded to BOPTEST later as testcases. I was wondering if there are any resources available that explain creating BOPTEST-compatible FMU using OpenModelica. I have the .mo files and the resource files available (which were used to compile as an FMU successfully using OPTIMICA and successfully ran on BOPTEST). 

I tried to compile the FMU using OMedit, and the compiled FMU failed to run on BOPTEST out of the box. I think the reason was that the resource files were not in the appropriate places.

I also need to switch to a script-based workflow, and leave OMEdit, so any suggestions on using a different tool are welcome.

**Comment by EttoreZ** [comment]
hey @jmythms. Going point by point:


> I'm trying to change my OPTIMICA-based FMU creation workflow to use OpenModelica instead. The FMUs will be uploaded to BOPTEST later as testcases. I was wondering if there are any resources available that explain creating BOPTEST-compatible FMU using OpenModelica. I have the .mo files and the resource files available (which were used to compile as an FMU successfully using OPTIMICA and successfully ran on BOPTEST).
> 
The BOPTEST team is working on implementing OpenModelica as a native compiler option for test cases. You can check branch [issue422](https://github.com/ibpsa/project1-boptest/blob/issue422_om_msl4/parsing/parser.py) `parser.py` script to see how the FMU is compiled with OpenModelica. In a future BOPTEST release OpenModelica will be made available as compiler option.

> I tried to compile the FMU using OMedit, and the compiled FMU failed to run on BOPTEST out of the box. I think the reason was that the resource files were not in the appropriate places.
> 
BOPTEST uses pyfmi to run FMUs, ideally OpenModelica FMUs should run out of the box, but that is not necessarily the case. I would test the FMU stand alone running in OpenModelica and then check the resource file structure.

> I also need to switch to a script-based workflow, and leave OMEdit, so any suggestions on using a different tool are welcome.
>
OpenModelica has a scripting [API](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/scripting_api.html) that can be used to create models, compile, create FMUs, simulate, plot, etc...

**Reply by jmythms** [reply]
Appreciate it! Will report back after trying these out!

**Comment by dhblum** [comment]
@EttoreZ Thanks for replying.  When you say "I would test the FMU stand alone running in OpenModelica and then check the resource file structure" do you mean "I would test the FMU stand lone running in pyfmi?" (instead of running in OpenModelica?)

**Reply by EttoreZ** [reply]
I am sorry about the confusion. I meant I would re-import the FMU in OpenModelica and test it there, then with pyfmi and then in BOPTEST.

**Comment by jmythms** [comment]
Seemed like one of the major issues was also that spawn path was not set ok. @kbenne helped me fix that and progress further. I'm currently hitting this error message, and we think it may be an OpenModelica solver issue:

```
While solving non-linear system an assertion failed at time 5.59494e-28.
DDASSL had repeated error test failures on the last attempted step.
can't continue. time = 0.000000
Homotopy solver total pivot: Matrix (nearly) singular at time 0.000000.
Homotopy solver total pivot: Matrix (nearly) singular at time 0.000000.
under-determined linear system not solvable!
Homotopy solver total pivot: Matrix (nearly) singular at time 0.000000.
under-determined linear system not solvable!
Homotopy solver total pivot: Matrix (nearly) singular at time 0.000000.
under-determined linear system not solvable!
The following assertion has been violated at time 0.000000 ((sinSpeDX1.eva.m >= 0.0)) --> "Variable violating min constraint: 0.0 <= sinSpeDX1.eva.m, has value: -3.17808e-43"
model terminate | Integrator failed. | Simulation terminated at time 5.20472e-28
### STATISTICS ###
0.000 Building.building: Warning from EnergyPlus: The resource referenced by LifeCycleCost:UsePriceEscalation= "U.S. AVG COMMERCIAL-ELECTRICITY" has no energy cost. 
0.000 Building.building: Warning from EnergyPlus: ... It is likely that the wrong resource is used. The resource should match the meter used in Utility:Tariff. 
0.000 Building.building: Warning from EnergyPlus: The resource referenced by LifeCycleCost:UsePriceEscalation= "U.S. AVG COMMERCIAL-DISTILLATE OIL" has no energy cost. 
0.000 Building.building: Warning from EnergyPlus: ... It is likely that the wrong resource is used. The resource should match the meter used in Utility:Tariff. 
0.000 Building.building: Warning from EnergyPlus: The resource referenced by LifeCycleCost:UsePriceEscalation= "U.S. AVG COMMERCIAL-RESIDUAL OIL" has no energy cost. 
0.000 Building.building: Warning from EnergyPlus: ... It is likely that the wrong resource is used. The resource should match the meter used in Utility:Tariff. 
0.000 Building.building: Warning from EnergyPlus: The resource referenced by LifeCycleCost:UsePriceEscalation= "U.S. AVG COMMERCIAL-NATURAL GAS" has no energy cost. 
0.000 Building.building: Warning from EnergyPlus: ... It is likely that the wrong resource is used. The resource should match the meter used in Utility:Tariff. 
0.000 Building.building: Warning from EnergyPlus: The resource referenced by LifeCycleCost:UsePriceEscalation= "U.S. AVG COMMERCIAL-COAL" has no energy cost. 
0.000 Building.building: Warning from EnergyPlus: ... It is likely that the wrong resource is used. The resource should match the meter used in Utility:Tariff. 
EnergyPlus Starting EnergyPlus, Version 9.6.0-316f894e91 (No OpenGL) (No OpenGL) (No OpenGL), YMD=2024.10.14 12:55 Adjusting Air System Sizing Adjusting Standard 62.1 Ventilation Sizing Initializing Simulation Reporting Surfaces Beginning Primary Simulation Initializing New Environment Parameters Warming up {1} Warming up {2} Warming up {3} Warming up {4} Warming up {5} Warming up {6} Warming up {7} Warming up {8} Starting Simulation at 01/01/2017 for SPAWN-RUNPERIOD DASKR-- AT T (=R1) AND STEPSIZE H (=R2) THE In above, R1 = 5.2047193883507E-28 R2 = 3.1384126340985E-29 DASKR-- ERROR TEST FAILED REPEATEDLY OR WITH ABS(H)=HMIN Writing tabular output file results using HTML format. Computing Life Cycle Costs and Reporting Writing final SQL reports EnergyPlus Run Time=00hr 00min 0.99sec 
EnergyPlus Completed Successfully. 
``` 

I'm attaching the case I am trying to simulate, it is the building.mo in the Heatpump_singleSpeed directory. 
[370.zip](https://github.com/user-attachments/files/17367241/370.zip) Wondering if you can take a look, @EttoreZ

**Reply by EttoreZ** [reply]
hey @jmythms. The error seems to be an initialization issue on the evaporator model of the cooling coil `Buildings.Fluid.HeatExchangers.DXCoils.BaseClasses.Evaporation` water mass accumulated ` Modelica.SIunits.Mass m(nominal=-5000*1400/2257E3, start=0, fixed=true)` . For some reason OpenModelica can't find a solution. To see if that is actually the problem I would initialize the simulation with a different starting value of `m` close to zero, but not zero. Another way is to start the simulation in a different day and see if the problem persists (the initial value of `m` depends on external conditions ) . Unfortunately I have no definitive solution for this, as mentioned before, OpenModelica has some issues with FMU generation. You can find a minimum error example and report it to [OpenModelica](https://github.com/OpenModelica/OpenModelica/issues).

**Reply by jmythms** [reply]
Do you think there will be any advantages to updating to the latest Buildings library? Or do you expect that OpenModelica is more of the reason for it? We are currently using Buildings Library v9.1.0.

**Reply by EttoreZ** [reply]
Updating from Buildings 9.1. to a more recent version can help. However, 9.1 already uses Modelica 4.0 if I memory serves me right and that is where a lot of changes were made. You can also try using the nightly build of OpenModelica instead of the stable to see if they made any useful changes very recently.

**Reply by jmythms** [reply]
Thank you for the advice. Exploring our options now.
